oeps/best-practices/oep-0066-bp-authorization.rst

```diff
@@ -0,0 +1,433 @@
+           'PUT': ['%(app_label)s.change_%(model_name)s'],
+           'PATCH': ['%(app_label)s.change_%(model_name)s'],
+           'DELETE': ['%(app_label)s.delete_%(model_name)s'],
+       }
+
+If additional information about the session is needed beyond the user's
+identity in order to make a permission decision (for example, if an action
+should only be allowed if the client has been granted a particular OAuth
+scope, as outlined in :doc:`OEP 4 </architectural-decisions/oep-0004-arch-oauth-scopes>`), 
+then a custom `BasePermission`_ subclass can
+be implemented which both consults the Django authorization API and makes the
+necessary checks against the session or other properties of the request
+object.
+
+In order to filter the querysets used to generate list responses to only
+include objects appropriate for the users permissions, an appropriate filter
+class should also be set.  A generic implementation using the library
+proposed above for mapping permissions to Q objects might look as follows:
+
+.. code-block:: python
+
+    from qpermissions import perms
+
+    class DjangoPermissionRulesFilter(BaseFilterBackend):
+
+        perm_format = '%(app_label)s.view_%(model_name)s'
+
+        def filter_queryset(self, request, queryset, view):
+            user = request.user
+            model_cls = queryset.model
+            kwargs = {
+                'app_label': model_cls._meta.app_label,
+                'model_name': get_model_name(model_cls)
+            }
+            permission = self.perm_format % kwargs
+            if permission not in perms:
+                return queryset
+            return queryset.filter(perms[permission](user))
+
+Such a class would be used in a view's `filter_backends`_ attribute or
+could be used by default for all view classes which don't override it.
+
+.. _BasePermission: https://www.django-rest-framework.org/api-guide/permissions/#custom-permissions
+.. _filter_backends: https://www.django-rest-framework.org/api-guide/filtering/#setting-filter-backends
+
+Systems/Protocols Overview
+***************************
+The following systems/protocols are currently used in the Open edX ecosystem 
+to grant users different levels of access. Each system/protocol is used in different 
+ways. A user's authz level is determined based on a combination of these systems/protocols. 
+It is important to note that it is the interplay of these systems/protocols that 
+determines whether a user has permissions for a given operation, not necessarily a single 
+system.
+
+Open edX Authorization Systems Diagram
+
+.. image:: oep-0066/Open_edX_Authorization.png
+   :alt: A diagram that shows the different systems/protocols used in Open edX to control authorization. The information in the diagram is also in the Open edX Authorization Systems Table (linked to in this document).
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   oep-0066/Open_edX_Authorization_Systems_Table.rst
+
+django Admin (auth_permission)
```

Comment:
is there a reason `django` is lowercase in this section but uppercase previously?